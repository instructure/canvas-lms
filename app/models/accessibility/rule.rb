# frozen_string_literal: true

#
# Copyright (C) 2025 - present Instructure, Inc.
#
# This file is part of Canvas.
#
# Canvas is free software: you can redistribute it and/or modify it under
# the terms of the GNU Affero General Public License as published by the Free
# Software Foundation, version 3 of the License.
#
# Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
# A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
# details.
#
# You should have received a copy of the GNU Affero General Public License along
# with this program. If not, see <http://www.gnu.org/licenses/>.

module Accessibility
  class Rule
    class << self
      attr_accessor :id, :link

      def registry
        rules = [
          Accessibility::Rules::AdjacentLinksRule,
          Accessibility::Rules::HeadingsSequenceRule,
          Accessibility::Rules::HeadingsStartAtH2Rule,
          Accessibility::Rules::ImgAltFilenameRule,
          Accessibility::Rules::ImgAltLengthRule,
          Accessibility::Rules::ImgAltRule,
          Accessibility::Rules::LargeTextContrastRule,
          Accessibility::Rules::ListStructureRule,
          Accessibility::Rules::ParagraphsForHeadingsRule,
          Accessibility::Rules::SmallTextContrastRule,
          Accessibility::Rules::TableCaptionRule,
          Accessibility::Rules::TableHeaderRule,
          Accessibility::Rules::TableHeaderScopeRule
        ]

        rules.index_by(&:id)
      end

      def pdf_registry
        [
          Accessibility::Rules::HasLangEntryRule,
        ]
      end

      # Tests if an element passes this accessibility rule
      # @param elem [Nokogiri::XML::Element] The element to test
      # @return [String] nil if the element passes, the validation problem
      # if there's an issue
      def test(elem)
        raise NotImplementedError, "#{self} must implement/override test"
      end

      # Gets a form definition for correcting the issue
      # @param _elem [Nokogiri::XML::Element] The element in the content to fix
      # @return [FormField] Form field for correcting the issue
      def form(_elem)
        {}
      end

      # Generates a fix for the issue in the content, to be used in the
      # correction form
      # @param _elem [Nokogiri::XML::Element] The element in the content to fix
      # @return [String] The generated fix, or nil if generation is not supported
      # @throws [StandardError] if the fix cannot be generated for a reason
      def generate_fix(_elem)
        nil
      end

      # Fixes the issue in content, changing it in-place
      # @param _elem [Nokogiri::XML::Element] The element in the content to fix
      # @param _value [String] The value received back from the correction form
      # @throws [StandardError] if the fix cannot be applied for a reason
      # @throws [NotImplementedError] if not implemented in a rule
      # @return [String] Potential error message if the fix fails, or nil if it
      # succeeds
      def fix!(_elem, _value)
        raise NotImplementedError, "#{self} must implement fix"
      end

      # Gets the name of this rule
      # @return [String] The name of this rule
      def display_name
        raise NotImplementedError, "#{self} must implement display_name"
      end

      # Gets the message for users about this rule
      # @return [String] The message about this issue
      def message
        raise NotImplementedError, "#{self} must implement message"
      end

      # Gets the explanation of why this issue is important
      # @return [String] The explanation of the issue
      def why
        raise NotImplementedError, "#{self} must implement/override why"
      end
    end
  end
end
