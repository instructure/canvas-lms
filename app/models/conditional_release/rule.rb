# frozen_string_literal: true

#
# Copyright (C) 2020 - present Instructure, Inc.
#
# This file is part of Canvas.
#
# Canvas is free software: you can redistribute it and/or modify it under
# the terms of the GNU Affero General Public License as published by the Free
# Software Foundation, version 3 of the License.
#
# Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
# A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
# details.
#
# You should have received a copy of the GNU Affero General Public License along
# with this program. If not, see <http://www.gnu.org/licenses/>.

module ConditionalRelease
  class Rule < ActiveRecord::Base
    include Deletion

    validates :course_id, presence: true
    validates :trigger_assignment_id, presence: true
    validate :trigger_assignment_in_same_course
    belongs_to :trigger_assignment, :class_name => "Assignment"

    belongs_to :course
    belongs_to :root_account, :class_name => "Account"
    has_many :scoring_ranges, -> { active.order(position: :asc) }, inverse_of: :rule, dependent: :destroy
    has_many :assignment_sets, -> { active }, through: :scoring_ranges
    has_many :assignment_set_associations, -> { active.order(position: :asc) }, through: :scoring_ranges
    accepts_nested_attributes_for :scoring_ranges, allow_destroy: true

    after_save :clear_caches

    before_create :set_root_account_id
    def set_root_account_id
      self.root_account_id ||= course.root_account_id
    end

    def trigger_assignment_in_same_course
      if trigger_assignment_id_changed? && trigger_assignment&.context_id != course_id
        errors.add(:trigger_assignment_id, "invalid trigger assignment")
      end
    end

    scope :with_assignments, -> do
      having_assignments = joins(Rule.preload_associations).group(Arel.sql("conditional_release_rules.id"))
      preload(Rule.preload_associations).where(id: having_assignments.pluck(:id))
    end

    def self.preload_associations
      { scoring_ranges: { assignment_sets: :assignment_set_associations } }
    end

    def self.includes_for_json
      {
        scoring_ranges: {
          include: {
            assignment_sets: {
              include: {assignment_set_associations: {except: [:root_account_id, :deleted_at]}},
              except: [:root_account_id, :deleted_at]
            }
          },
          except: [:root_account_id, :deleted_at]
        }
      }
    end

    def assignment_sets_for_score(score)
      AssignmentSet.active.where(scoring_range: scoring_ranges.for_score(score))
    end

    def clear_caches
      self.class.connection.after_transaction_commit do
        self.trigger_assignment.clear_cache_key(:conditional_release)
        self.course.clear_cache_key(:conditional_release)
      end
    end

    def self.is_trigger_assignment?(assignment)
      # i'm only using the cache key currently for this one case but i figure it can be extended to handle caching around all rule data fetching
      RequestCache.cache('conditional_release_is_trigger', assignment) do
        Rails.cache.fetch_with_batched_keys('conditional_release_is_trigger', batch_object: assignment, batched_keys: :conditional_release) do
          assignment.shard.activate { self.active.where(:trigger_assignment_id => assignment).exists? }
        end
      end
    end
  end
end
