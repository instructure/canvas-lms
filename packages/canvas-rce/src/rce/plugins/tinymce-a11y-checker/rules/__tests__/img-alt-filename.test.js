/*
 * Copyright (C) 2023 - present Instructure, Inc.
 *
 * This file is part of Canvas.
 *
 * Canvas is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, version 3 of the License.
 *
 * Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import rule from '../img-alt-filename'

let el

beforeEach(() => {
  el = document.createElement('img')
  el.setAttribute('src', '/file.txt')
})

describe('test', () => {
  test('returns true if alt text is missing', () => {
    expect(rule.test(el)).toBeTruthy()
  })

  test('returns true if not img tag', () => {
    expect(rule.test(el)).toBeTruthy()
  })

  test('returns true if alt text is empty', () => {
    el.setAttribute('alt', '')
    expect(rule.test(el)).toBeTruthy()
  })

  test('returns true if alt text is not filename', () => {
    el.setAttribute('alt', 'some text')
    expect(rule.test(el)).toBeTruthy()
  })

  test('returns false if alt text is filename', () => {
    el.setAttribute('alt', 'file.txt')
    expect(rule.test(el)).toBeFalsy()
  })

  test('returns false if alt text is filename with blank spaces', () => {
    el.setAttribute('alt', 'file with blank.txt')
    expect(rule.test(el)).toBeFalsy()
  })
})

describe('data', () => {
  test('returns alt text', () => {
    el.setAttribute('alt', 'some text')
    expect(rule.data(el).alt).toBe('some text')
  })

  test('returns empty alt text if no alt attribute', () => {
    expect(rule.data(el).alt).toBe('')
  })

  test('returns decorative true if el has empty alt attribute', () => {
    el.setAttribute('alt', '')
    expect(rule.data(el).decorative).toBeTruthy()
  })
})

describe('form', () => {
  test('returns the proper object', () => {
    expect(rule.form()).toMatchSnapshot()
  })
})

describe('update', () => {
  test('returns same element', () => {
    expect(rule.update(el, {})).toBe(el)
  })

  test("sets alt text to empty and role to 'presentation' if decorative", () => {
    rule.update(el, {decorative: true})
    expect(el.getAttribute('alt')).toBe('')
    expect(el.getAttribute('role')).toBe('presentation')
  })

  test('sets alt text and removes role if not decorative', () => {
    el.setAttribute('alt', '')
    el.setAttribute('role', 'presentation')
    rule.update(el, {decorative: false, alt: 'some text'})
    expect(el.getAttribute('alt')).toBe('some text')
    expect(el.hasAttribute('role')).toBeFalsy()
  })

  test('changes alt text if not decorative', () => {
    const text = 'this is my text'
    el.setAttribute('alt', 'thisismy.txt')
    expect(rule.update(el, {alt: text, decorative: false}).getAttribute('alt')).toBe(text)
  })

  test('removes role if not decorative', () => {
    const text = 'this is my text'
    rule.update(el, {alt: text, decorative: false})
    expect(el.hasAttribute('role')).toBeFalsy()
    expect(el.getAttribute('alt')).toBe(text)
  })

  test('sets alt text to empty and role if decorative', () => {
    rule.update(el, {decorative: true})
    expect(el.getAttribute('role')).toBe('presentation')
    expect(el.getAttribute('alt')).toBe('')
  })
})

describe('message', () => {
  test('returns the proper message', () => {
    expect(rule.message()).toMatchSnapshot()
  })
})

describe('why', () => {
  test('returns the proper why message', () => {
    expect(rule.why()).toMatchSnapshot()
  })
})

describe('linkText', () => {
  test('returns the proper why message', () => {
    expect(rule.linkText()).toMatchSnapshot()
  })
})
