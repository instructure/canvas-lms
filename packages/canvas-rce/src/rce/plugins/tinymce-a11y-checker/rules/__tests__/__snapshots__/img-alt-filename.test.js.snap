// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`form returns the proper object 1`] = `
Array [
  Object {
    "dataKey": "alt",
    "disabledIf": [Function],
    "label": "Change alt text",
  },
  Object {
    "checkbox": true,
    "dataKey": "decorative",
    "label": "Decorative image",
  },
]
`;

exports[`linkText returns the proper why message 1`] = `"Learn more about using filenames as alt text"`;

exports[`message returns the proper message 1`] = `"Image filenames should not be used as the alt attribute describing the image content."`;

exports[`why returns the proper why message 1`] = `"Screen readers cannot determine what is displayed in an image without alternative text, and filenames are often meaningless strings of numbers and letters that do not describe the context or meaning."`;
